src/libxtp/davidsonsolver.cc

```diff
@@ -165,21 +201,116 @@ Eigen::ArrayXi DavidsonSolver::argsort(const Eigen::VectorXd &V) const {
+        guess(idx(j), j) = 1.0;
+      }
+      break;
+
+    case TYPE::HAM:
+      /* Initialize the guess eigenvector so that they 'target' the lowest
+       * positive diagonal elements */
+        int ind0 = d.size()/2;
+        int shift = size_initial_guess/4;
+      for (int j = 0; j < size_initial_guess; j++) {
+        guess(idx(ind0+j), j) = 1.0;
+      }
+      break;
+    }
   return guess;
 }
 
-Eigen::VectorXd DavidsonSolver::dpr_correction(const Eigen::VectorXd &r,
+DavidsonSolver::RitzEigenPair DavidsonSolver::getRitz (
+    const DavidsonSolver::ProjectedSpace &proj, int size_update) const {
+
+  DavidsonSolver::RitzEigenPair rep;
+  Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> es(proj.T);
+  rep.lambda = es.eigenvalues();
+  rep.U = es.eigenvectors();
+      
+  rep.q = proj.V * rep.U;  // Ritz vectors 
+  rep.res = proj.AV * rep.U - rep.q * rep.lambda.asDiagonal();  // residues
+  rep.res_norm = rep.res.colwise().norm(); // reisdues norms
+
+  return rep;
+
+}
+
+DavidsonSolver::ProjectedSpace DavidsonSolver::initProjectedSpace(Eigen::VectorXd &Adiag, 
+                                                  int size_initial_guess) const {
+  DavidsonSolver::ProjectedSpace proj; 
+  proj.V = DavidsonSolver::setupInitialEigenvectors(Adiag, size_initial_guess);
+  proj.search_space = proj.V.cols();
+  return proj;
+}
+
+int DavidsonSolver::extendProjection(Eigen::VectorXd &Adiag, 
+    DavidsonSolver::RitzEigenPair &rep,  DavidsonSolver::ProjectedSpace &proj, 
+    std::vector<bool> &root_converged, int size_update) {
+
+  int nupdate = 0;
+  for (int j = 0; j < size_update; j++) {
+    // skip the root that have already converged
+    if (this->_matrix_type == TYPE::SYMM) {
+      if (root_converged[j]) {
+        continue;
+      }
+    }
+    nupdate++;
+
+    // residue vector
+    Eigen::VectorXd w =
+        computeCorrectionVector(Adiag, rep.q.col(j), rep.lambda(j), rep.res.col(j));
+
+    // append the correction vector to the search space
+    proj.V.conservativeResize(Eigen::NoChange, proj.V.cols() + 1);
+    proj.V.rightCols<1>() = w.normalized();
+
+    // track converged root
+    root_converged[j] = (rep.res_norm[j] < _tol);
+  }
+
+  proj.search_space = proj.V.cols();
+
+  return nupdate;
+}
+
+Eigen::MatrixXd DavidsonSolver::extract_eigenvectors(const Eigen::MatrixXd &V, 
+                                                     const Eigen::ArrayXi &idx) 
+                                                     const {
+  Eigen::MatrixXd W = Eigen::MatrixXd::Zero(V.rows(),idx.size());
+  for (int i=0; i < idx.size(); i++) {
+    W.col(i) = V.col(idx(i));
+  }
+  return W;
+}
+
+Eigen::VectorXd DavidsonSolver::computeCorrectionVector(
+    const Eigen::VectorXd &Adiag, const Eigen::VectorXd &qj, double lambdaj,
+    const Eigen::VectorXd &Aqj) const {
+  Eigen::VectorXd w;
+  // compute correction vector
+  switch (this->_davidson_correction) {
+    case CORR::DPR:
+      w = dpr(Aqj, Adiag, lambdaj);
```

Comment:
you can just return in the switch case than you do not need w, or the break statements. So is fine the other way is shorter

Reply:
shorter is better